expression/builtin_string.go

```diff
@@ -3976,3 +3976,84 @@ func (b *builtinWeightStringSig) evalString(row chunk.Row) (string, bool, error)
 	}
 	return string(ctor.Key(str)), false, nil
 }
+
+type translateFunctionClass struct {
+	baseFunctionClass
+}
+
+// getFunction sets translate built-in function signature.
+// The syntax of translate in Oracle is 'TRANSLATE(expr, from_string, to_string)'.
+func (c *translateFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString, types.ETString, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+	argType := args[0].GetType()
+	bf.tp.Flen = argType.Flen
+	SetBinFlagOrBinStr(argType, bf.tp)
+	sig := &builtinTranslateSig{bf}
+	return sig, nil
+}
+
+type builtinTranslateSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinTranslateSig) Clone() builtinFunc {
+	newSig := &builtinTranslateSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalString evals a builtinTranslateSig, corresponding to translate(srcStr, fromStr, toStr)
+// See https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions196.htm
+func (b *builtinTranslateSig) evalString(row chunk.Row) (d string, isNull bool, err error) {
+	var (
+		srcStr, fromStr, toStr     string
+		isFromStrNull, isToStrNull bool
+		lenFrom, lenTo, minLen     int
+		tgt                        strings.Builder
+	)
+	srcStr, isNull, err = b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return d, isNull, err
+	}
+	fromStr, isFromStrNull, err = b.args[1].EvalString(b.ctx, row)
+	if isFromStrNull || err != nil {
+		return d, isNull, err
+	}
+	toStr, isToStrNull, err = b.args[2].EvalString(b.ctx, row)
+	if isToStrNull || err != nil {
+		return d, isNull, err
+	}
+	if srcStr == "" || fromStr == "" || toStr == "" {
+		return d, true, err
+	}
```

Comment:
:-1:

i question the reason to replicate Oracle's behavior of treating empty string as NULL here.

the `TRANSLATE()` function is not exclusive to Oracle. this function also exists in [PostgreSQL](https://www.postgresql.org/docs/current/functions-string.html#FUNCTIONS-STRING-TRANSLATE), [Microsoft SQL Server](https://docs.microsoft.com/en-us/sql/t-sql/functions/translate-transact-sql) and [IBM Db2](https://www.ibm.com/support/producthub/db2/docs/content/SSEPGG_11.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000862.html)*, where for all of them empty strings are distinct from NULL (unless using Oracle compatibility mode).

<details><summary>In PostgreSQL, using empty string as fromStr or toStr is the same as removing these 3 lines.</summary>

```console
# select translate('xyz', '', 'pqr');
xyz

# select translate('abc', 'aceg', '');
b

# select translate('fgh', '', '');
fgh
```

</details>

<details><summary>In MS SQL, the fromStr and toStr must have same length; but both being empty still won't give us NULL</summary>

```console
# select translate('xyz', '', 'pqr');
EREQUEST The second and third arguments of the TRANSLATE built-in function must contain an equal number of characters.

# select translate('abc', 'aceg', '');
EREQUEST The second and third arguments of the TRANSLATE built-in function must contain an equal number of characters.

# select translate('fgh', '', '');
fgh
```

</details>

(*: Db2 flips the order of fromStr and toStr. i don't have a Db2 to test.)

if we do want to copy Oracle's behavior, it should be an `@@SQL_MODE` switch that turns empty string into NULL, or instruct the customer to insert a [`NULLIF(fromStr, '')` call](https://dev.mysql.com/doc/refman/8.0/en/flow-control-functions.html#function_nullif), not adding a special case to the TRANSLATE function.